tests: Add the integration test for live migration

likebreath opened the pull request:
This test exercises the local live-migration between two Cloud
Hypervisor VMs on the same host. It ensures the following behaviors:
1. The source VM is up and functional (including various virtio-devices
are working properly);
2. The 'send-migration' and 'receive-migration' command finished
successfully;
3. The source VM terminated gracefully after live migration;
4. The destination VM is functional (including various virtio-devices
are working properly) after live migration.

Note: This test does not use vsock as we can't create two identical
vsock on the same host.

Fixes: #2965

Signed-off-by: Bo Chen <chen.bo@intel.com>

likebreath commented:
@sboeuf This is a WIP draft with the skeleton of local live migration. I am thinking to add a bit more check in the guest, say adding tag to the kernel cmdline and verify it is consistent on the destination side (like what we did for the vfio test). Please let me know if you have any comments or questions. Thank you!

sboeuf commented:
@likebreath 
>This is a WIP draft with the skeleton of local live migration. I am thinking to add a bit more check in the guest, say adding tag to the kernel cmdline and verify it is consistent on the destination side (like what we did for the vfio test). Please let me know if you have any comments or questions. Thank you!

That's a good start, I agree adding a TAG brings some added value, the same way you could also create an environment variable or a file and check the value on the destination is the one that has been set by the source.
I would also look at `test_snapshot_restore` (maybe factorize a bit to be able to reuse some parts) so that we perform the same kind of testing regarding the different types of VIRTIO devices. We want to check that both block, console, rng, net, vsock work well after the migration is done.
About vhost-user I'd suggest we add a dedicated live migration test as well (relying on OVS DPDK), but that can definitely happen in a separate PR.

likebreath commented:
@rbradford  I addressed the comment from @sboeuf about adding the checks on virtio-devices both before and post live-migration. PTAL and let me know if you have any comments. 

The CI reported this test is failing on the aarch64 platform, and the reason is the ssh command failed on the destination VM (after live-migration) with an error `Connection(Os { code: 113, kind: Other, message: "No route to host" })`. Any thoughts about what's special for the aarch64 platform (given the same test works fine on the x86_64 platform)? @michael2012z @MrXinWang 

The CI log is [here](https://cloud-hypervisor-jenkins.westus.cloudapp.azure.com/blue/organizations/jenkins/cloud-hypervisor/detail/PR-2978/2/pipeline) and the error message is below:

```
==== Start ssh command output (FAILED) ====

command="grep -c processor /proc/cpuinfo"
auth="PasswordAuth {
    username: "cloud",
    password: "cloud123",
}"

ip="192.168.21.2"
output=""
error="Connection(Os { code: 113, kind: Other, message: "No route to host" })"

==== End ssh command outout ====
```

Thanks a lot.

MrXinWang commented:
@likebreath Debugging with @michael2012z now :)) From the test I did locally, it seems that 2 things are strange:

-  I was following the doc from https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/docs/live_migration.md, where it says `When the above commands completed, the source VM should be successfully migrated to the destination VM. Now the destination VM is running while the source VM is paused.`. In the test I did locally, after the migration, the source VM is killed.
- The destination is migrated, however all interrupts from PCI devices were not working. I assume the migrated GICITS is not (at least not fully) functional, since 2 configuration tables of GICITS is stored in the memory, and maybe dirty page will affect that?

likebreath commented:
@rbradford @sboeuf The PR is ready for review. PTAL. Thanks.

I am working on adding some more tests on live-migration. Currently looking at NUMA/memory-zones and then ovs-dpdk (planning to reuse/refactor what we have from "test_guest_numa_nodes" and "test_ovs_dpdk").  

The simple 'memory-zone' test I did was working fine, but migrating a VM with the configs from "test_guest_numa_nodes" failed with error "oom". Not sure whether it is caused by using "virti-mem" as hotplug method, or simply the guest RAM are too big. I will dig deeper and ask your help when needed.

MrXinWang commented on tests/integration.rs:
```diff
@@ -6772,4 +6772,216 @@ mod tests {
+            }
+
+            // Check the source vm has been terminated successful (give it '1s' to settle)
+            thread::sleep(std::time::Duration::new(1, 0));
```
Can we extend the time here to `3s`? In my local test on arm64 machines, 1s seems to be not enough and sometimes it will give:
```
thread 'main' panicked at 'Test failed: source VM was not terminated successfully.', tests/integration.rs:7001:17
```

I tried to extend here to 3s and it will pass on arm.
```
running 1 test
test tests::live_migration::test_live_migration ... ok
```

likebreath replied:
Cool. This is a good finding. I will update the PR accordingly.

likebreath replied:
@MrXinWang In addition to what you suggested above, I also made a few more changes: 

1. Removed the "x86_64" feature gate for the "live-migration" test mod, so that we can run it on the aarch64 worker once your PR is landed; 

2. Added a "1s" sleep between the "receive-migration" and "send-migration", which should prevent the "random" failures of `VmSendMigration(VmSendMigration(MigrateSend(Error connecting to UNIX socket: No such file or directory (os error 2))))`.

MrXinWang replied:
Perfect! Great thanks @likebreath! I will do the refactor that @sboeuf  suggested in https://github.com/cloud-hypervisor/cloud-hypervisor/pull/3049 and take care of the rest on arm64 :))

likebreath replied:
Yes, I am also following that PR and agrees with the proposal. Thanks again for your hard work and prompt response.

likebreath commented:
> The simple 'memory-zone' test I did was working fine, but migrating a VM with the configs from "test_guest_numa_nodes" failed with error "oom". Not sure whether it is caused by using "virti-mem" as hotplug method, or simply the guest RAM are too big. I will dig deeper and ask your help when needed.

Alright. There were two issues that stop me from reusing `test_guest_numa_nodes` for adding a new "live-migration" test:
1. It is using "virtio-mem" for memory hotplug with memory-zones, while we are still missing snapshot/restore for "virtio-mem";
2. It is allocating too much guest RAM that can trigger OOM while migration.

I believe I will end up writing a separate test (based on `test_guest_numa_nodes`) that uses "memory-zones" and "numa nodes" but with less guest RAM and without memory-hotplug. Of course, it will happen in a separate PR.

sboeuf reviewed (approved):
@likebreath @rbradford are we waiting for something else from this PR. It looks good and I think is ready to be merged.

rbradford commented on scripts/run_integration_tests_live_migration.sh:
```diff
@@ -0,0 +1,87 @@
+if [ ! -f "$VMLINUX_IMAGE" ]; then
+    SRCDIR=$PWD
+    pushd $WORKLOADS_DIR
+    time git clone --depth 1 "https://github.com/cloud-hypervisor/linux.git" -b "ch-5.13-rc5" $LINUX_CUSTOM_DIR
```
This needs updating ^

likebreath replied:
Good catch. Updated to use "ch-5.14".
